crates/ide-completion/src/completions/flyimport.rs

```diff
@@ -377,13 +396,29 @@ fn import_assets_for_path(
         &ctx.sema,
         ctx.token.parent()?,
     )?;
+
     if fuzzy_name_length < 3 {
         cov_mark::hit!(flyimport_exact_on_short_path);
         assets_for_path.path_fuzzy_name_to_exact(false);
     }
+
     Some(assets_for_path)
 }
 
+fn is_expected_type(ctx: &CompletionContext<'_>, located_import: &LocatedImport) -> bool {
```

Comment:
This is the wrong way to approach it, our `CompletionItem` structure has a `type_match` field which is what we need to populate here. According to that, we are later going to sort the completions already. The `render_resolution_with_import(RenderContext::new(ctx), path_ctx, import)` call already does this for locals being completed in `render_resolution_path`. Now, the reason this therefore doesn't work here in flyimport already is that we are completin importable things which are never locals! See
https://github.com/Veykril/rust-analyzer/blob/994df3d6a31d39f11600f30a6df0b744b13937c1/crates/ide-completion/src/render.rs#L337-L351

We should instead generalize that part I believe, making it work for more scopedefs, mainly
`ScopeDef::ImplSelfType`, `ScopeDef::AdtSelfType` and reklevant `ScopeDef::ModuleDef`. For `ModuleDef`, all 3 `Adt`s, `Function`, `Variant`, `Const`, `Static` and `BuiltinType` are relevant. (Some of those might not have a function allowing you to easily turn them into a hir version yet. It's fine to skip them in that case (or (ask about how) to implement them).

Reply:
Awesome that makes a lot more sense haha.

Reply:
I used `UPDATE_EXPECT=1` to see what the new code produces. It looks more or less correct to me. I am not sure why there are two completion items though. Is it because we can choose to either import or use the fully qualified name? Also should `type_match: `Some(CouldUnify)` actually be  `type_match: `Some(Exact)`? as this is the exact type we are expecting.
```rust   
// crates/ide-completion/src/render.rs
 #[test]
    fn sort_suggested_structs_by_types() {
        check(
            r#"
//- /lib.rs crate:dep

pub mod test_mod_b {
    pub struct Struct {}
}

//- /main.rs crate:main deps:dep

fn test(input: dep::test_mod_b::Struct) { }

fn main() {
    test(Struct$0);
}
"#,
            SymbolKind::Struct,
            expect![[r#"
                [
                    CompletionItem {
                        label: "Struct (use dep::test_mod_b::Struct)",
                        source_range: 67..73,
                        delete: 67..73,
                        insert: "Struct",
                        kind: SymbolKind(
                            Struct,
                        ),
                        lookup: "Struct",
                        detail: "Struct",
                        relevance: CompletionRelevance {
                            exact_name_match: false,
                            type_match: Some(
                                CouldUnify,
                            ),
                            is_local: false,
                            is_item_from_trait: false,
                            is_name_already_imported: false,
                            requires_import: false,
                            is_op_method: false,
                            is_private_editable: false,
                            postfix_match: None,
                            is_definite: false,
                        },
                    },
                    CompletionItem {
                        label: "dep::test_mod_b::Struct {…}",
                        source_range: 67..73,
                        delete: 67..73,
                        insert: "dep::test_mod_b::Struct {  }$0",
                        kind: SymbolKind(
                            Struct,
                        ),
                        lookup: "test_mod_b::Struct{}",
                        detail: "dep::test_mod_b::Struct {  }",
                        relevance: CompletionRelevance {
                            exact_name_match: false,
                            type_match: Some(
                                CouldUnify,
                            ),
                            is_local: false,
                            is_item_from_trait: false,
                            is_name_already_imported: false,
                            requires_import: false,
                            is_op_method: false,
                            is_private_editable: false,
                            postfix_match: None,
                            is_definite: false,
                        },
                        trigger_call_info: true,
                    },
                ]
            "#]],
        );
    }
```

Reply:
The qualified completion is from another code path, you can ignore that one. Exact doesn't work in this case, since imported things might have generics that we aren't filling out when constructing their type (since we don't know the concrete generics for them).